### PR TITLE
frontend/fix/#1165 added driver location deviation service

### PR DESCRIPTION
**Pull request by soumyajit2304**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Added driver deviation count in location update service

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (hkmangla) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:392]: Can we create a common function for this?
```
            JSONObject startLatLng = new JSONObject();
            startLatLng.put("lat", startLat);
            startLatLng.put("lon", startLng);

            JSONObject endLatLng = new JSONObject();
            endLatLng.put("lat", endLat);
            endLatLng.put("lon", endLng);
```
Something like this

```
getLatLongObject (lat, long) {
           JSONObject latLng = new JSONObject();
            latLng.put("lat", endLat);
            latLng.put("lon", endLng);
return latLng
}
```

**Review by hkmangla — DISMISSED** [review]

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (soumyajit2304) [reply Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java]: Done

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:392]: Change requestMethod to `GET` in purescript also switch the `/route` api methods to `GET` from `POST`

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:415]: here we can have check for just if `respCode == 200`, no need to compute in case of failure. Instead just add a firebase event for failure case event: `LS_ERROR_GETTING_ROUTE`.

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:433]: No need to store pathLatLng into SharedPref you can create a set and get function in this file which can be accessed by the class as this variable scope is within this class.

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:446]: remove extra line

Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:446]: remove extra line here

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:448]: make variables camel-case

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:462]: We have a function to update SharedPref variables in this file :- `updateStorage(k, v)` use that function here

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:480]: For fetching value from storage also create a get function and use it across

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:486]: isEmpty() will give NullPointerException if applied on a null value. Which is possible as the default value passed while getting it from storage is `null`. 

You can have checks in two ways:-

`Objects.isNull(waypoints)`

or 

`waypoints == null`

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:498]: add a null check for the above values before making the API call. Else API call will be failed, some on other places wherever variable is being fetched from storage

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:812]: why is wrapping this function with try/catch block required? Function itself has a try/catch block within

File: Frontend/ui-driver/src/Helpers/Storage.purs
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Helpers/Storage.purs:75]: Remove `RIDE_WAYPOINT` if not required

**Review by witcher-shailesh — CHANGES_REQUESTED** [review]
Add null check as well before making API calls where data is being fetched from storage

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (soumyajit2304) [reply Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:812]: The function `callDriverCurrentLocationAPI` throws a json exception. 
So wherever it is called it needs to be within try catch block.

Comment (witcher-shailesh) [reply Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:812]: Then handle the exception within the function, no need to add throws Exception in front of function

**Comment by soumyajit2304** [comment]
Done the required changes.

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:450]: If there's a try/catch block with the whole function. No need to throw JSONException off the function

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:460]: the threshold here is 30.0, can you make it accessible via local storage variable?

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:508]: No need to have separate try/catch block on this function, just remove throws exception from the function

File: Frontend/ui-driver/src/Services/APITypes.purs
Comment (witcher-shailesh) [inline Frontend/ui-driver/src/Services/APITypes.purs:280]: Make it as maybe type to have backward compatibility

**Comment by soumyajit2304** [comment]
> Added comments

Done

**Review by witcher-shailesh — APPROVED** [review]

File: Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java
Comment (witcher-shailesh) [inline Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java:424]: Until `:rideId/route` is merged and deployed to PROD. We have to keep a fallback where the API called should be `trip/route` 

So, In else can you handle the error code? If it's 404 NOT_FOUND we can call the `trip/route`

**Review by witcher-shailesh — CHANGES_REQUESTED** [review]

**Review by witcher-shailesh — APPROVED** [review]

**Pull request merged**
